File: C#/TollCalculator.cs
﻿using TollFeeCalculator;

public class TollCalculator
{
    const int DAILY_MAX_FEE = 60;

    /**
     * Calculate the total toll fee for one day
     *
     * @param vehicle - the vehicle
     * @param dates   - date and time of all passes on one day
     * @return - the total toll fee for that day
     */

    public int GetTollFee(Vehicle vehicle, DateTime[] dates)
    {
        int totalFee = 0;

        var sortedDates = GetDateTimeSpans(dates);

        foreach(var sortedDatesList in sortedDates)
        {
            int oldFee = 0;

            foreach(var dt in sortedDatesList)
            {
                int newFee = GetTollFee(dt, vehicle);
                if(newFee > oldFee)
                {
                    oldFee = newFee;
                }
            }
            
            totalFee += oldFee;
        }

        if (totalFee > DAILY_MAX_FEE) totalFee = DAILY_MAX_FEE;
        return totalFee;
    }

    /// <summary>
    /// Split the DateTimes into segments that are each one hour long.
    /// </summary>
    /// <param name="dates"></param>
    /// <returns></returns>
    public List<List<DateTime>> GetDateTimeSpans(DateTime[] dates)
    {
        List<List<DateTime>> dateTimes = new List<List<DateTime>>();
        for(int i = 0; i < dates.Length; i++)
        {
            // Add the first element
            if(i == 0)
            {
                dateTimes.Add(new List<DateTime>());
                dateTimes[dateTimes.Count - 1].Add(dates[i]);
                continue;
            }

            // Compare the current date time to the first element in the list. Has more than one hour passed? If so, start on the next list.
            if(DateTime.Compare(dates[i],
                dateTimes[dateTimes.Count - 1][0].AddHours(1)) > 0)
            {
                dateTimes.Add(new List<DateTime>());
                dateTimes[dateTimes.Count - 1].Add(dates[i]);
            }
            else
            {
                dateTimes[dateTimes.Count - 1].Add(dates[i]);
            }
        }

        return dateTimes;
    }

    private bool IsTollFreeVehicle(Vehicle vehicle)
    {
        if (vehicle == null) return false;
        String vehicleType = vehicle.GetVehicleType();
        return vehicleType.Equals(TollFreeVehicles.Motorbike.ToString()) ||
               vehicleType.Equals(TollFreeVehicles.Tractor.ToString()) ||
               vehicleType.Equals(TollFreeVehicles.Emergency.ToString()) ||
               vehicleType.Equals(TollFreeVehicles.Diplomat.ToString()) ||
               vehicleType.Equals(TollFreeVehicles.Foreign.ToString()) ||
               vehicleType.Equals(TollFreeVehicles.Military.ToString());
    }

    public int GetTollFee(DateTime date, Vehicle vehicle)
    {
        if (IsTollFreeDate(date) || IsTollFreeVehicle(vehicle)) return 0;

        int hour = date.Hour;
        int minute = date.Minute;

        if (hour == 6 && minute >= 0 && minute <= 29) return 8;
        else if (hour == 6 && minute >= 30 && minute <= 59) return 13;
        else if (hour == 7 && minute >= 0 && minute <= 59) return 18;
        else if (hour == 8 && minute >= 0 && minute <= 29) return 13;
        else if (hour == 8 && minute >= 30 && minute <= 59) return 8;
        else if (hour >= 9 && hour <= 14) return 8;
        else if (hour == 15 && minute >= 0 && minute <= 29) return 13;
        else if (hour == 15 && minute >= 29 || hour == 16 && minute <= 59) return 18;
        else if (hour == 17 && minute >= 0 && minute <= 59) return 13;
        else if (hour == 18 && minute >= 0 && minute <= 29) return 8;
        else return 0;
    }

    private Boolean IsTollFreeDate(DateTime date)
    {
        int year = date.Year;
        int month = date.Month;
        int day = date.Day;

        if (date.DayOfWeek == DayOfWeek.Saturday || date.DayOfWeek == DayOfWeek.Sunday) return true;

        if (year == 2013)
        {
            if (month == 1 && day == 1 ||
                month == 3 && (day == 28 || day == 29) ||
                month == 4 && (day == 1 || day == 30) ||
                month == 5 && (day == 1 || day == 8 || day == 9) ||
                month == 6 && (day == 5 || day == 6 || day == 21) ||
                month == 7 ||
                month == 11 && day == 1 ||
                month == 12 && (day == 24 || day == 25 || day == 26 || day == 31))
            {
                return true;
            }
        }
        return false;
    }

    private enum TollFreeVehicles
    {
        Motorbike = 0,
        Tractor = 1,
        Emergency = 2,
        Diplomat = 3,
        Foreign = 4,
        Military = 5
    }
}

File: Java/src/TollCalculator.java

<<<<<<< HEAD
=======
import vehicle.Vehicle;
>>>>>>> 36e9c27 (Fixing Code)
import java.time.DayOfWeek;
import java.time.Duration;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.Month;
import java.time.temporal.TemporalAdjusters;
import java.util.*;

public class TollCalculator {

  private static final Set<String> tollFreeVehicleTypes = Set.of("Motorbike", "Tractor", "Emergency", "Diplomat", "Foreign", "Military");

  /**
   * Calculate the total toll fee for one day
   *
   * @param vehicle - the vehicle
   * @param passes   - date and time of all passes on one day
   * @return - the total toll fee for that day
   */
  public int getTollFee(Vehicle vehicle, LocalDateTime... passes) {
    if (passes == null || passes.length == 0|| passes[0] == null) {
      throw new RuntimeException("No passes were given");
    }
    checkIfAllPassesIsTheSameDay(passes);
    LocalDateTime lastHighestIntervall = null;
    int lastHighestFee = 0;
    int totalFee = 0;
    for (LocalDateTime date : passes) {
      if (totalFee >= 60) {
        break;
      }

      int nextFee = getFeeOnePass(date, vehicle);

      if (lastHighestIntervall == null || Duration.between(lastHighestIntervall, date).toMinutes() > 60) {
        lastHighestFee = nextFee;
        lastHighestIntervall = date;
        totalFee += nextFee;

      } else {
        if (nextFee > lastHighestFee) {
          lastHighestFee = nextFee;
          lastHighestIntervall = date;
          totalFee += lastHighestFee;
        }
      }
    }
    if (totalFee > 60) totalFee = 60;
    return totalFee;
  }

  private void checkIfAllPassesIsTheSameDay(LocalDateTime... passes) {
    LocalDateTime firstPass = passes[0];
    LocalDate firstPassDate = firstPass.toLocalDate();

    for (int i = 1; i < passes.length; i++) {
      LocalDateTime currentPass = passes[i];
      LocalDate currentPassDate = currentPass.toLocalDate();

      if (!currentPassDate.equals(firstPassDate)) {
        throw new RuntimeException("Passes are on different days");
      }
    }
  }

  private boolean isTollFreeVehicle(Vehicle vehicle) {
    if (vehicle == null) return false;

    return tollFreeVehicleTypes.contains(vehicle.getType());
  }

  public int getFeeOnePass(final LocalDateTime date, Vehicle vehicle) {
    if(isTollFreeDate(date) || isTollFreeVehicle(vehicle)) return 0;
    int hour = date.getHour();
    int minute = date.getMinute();


    if (hour == 6 && minute >= 30 ) return 13;
    else if (hour == 7 ) return 18;
    else if (hour == 8 && minute <= 29) return 13;
    else if (hour == 15 && minute <= 29) return 13;
    else if (hour == 15 && minute >= 30 || hour == 16) return 18;
    else if (hour == 17) return 13;
    else return 8;
  }

  private Boolean isTollFreeDate(LocalDateTime date) {
    int year = date.getYear();

    DayOfWeek dayOfWeek = date.getDayOfWeek();
    if (dayOfWeek == DayOfWeek.SATURDAY || dayOfWeek == DayOfWeek.SUNDAY) return true;
    LocalDate EasterSunday = calculateEaster(year);

    Set<LocalDate> specialDates = new HashSet<>(Arrays.asList(
            LocalDate.of(year, Month.JANUARY, 1),
            LocalDate.of(year, Month.APRIL, 30),
            LocalDate.of(year, Month.MAY, 1),
            LocalDate.of(year, Month.JUNE, 5),
            LocalDate.of(year, Month.JUNE, 6),
            LocalDate.of(year, Month.DECEMBER, 24),
            LocalDate.of(year, Month.DECEMBER, 25),
            LocalDate.of(year, Month.DECEMBER, 26),
            LocalDate.of(year, Month.DECEMBER, 31),
            EasterSunday.minusDays(3), //Maundy Thursday
            EasterSunday.minusDays(2), //Good friday
            EasterSunday.plusDays(1), //Easter Monday
            EasterSunday.plusDays(39), //Ascension day
            calculateDateOfFridays(year, 24), //Midsummer
            calculateDateOfFridays(year, 43) //All Saints Day

    ));

    return specialDates.contains(date.toLocalDate());
  }

  private LocalDate calculateEaster(int year) {
    int a = year % 19;
    int b = year / 100;
    int c = year % 100;
    int d = b / 4;
    int e = b % 4;
    int f = (b + 8) / 25;
    int g = (b - f + 1) / 3;
    int h = (19 * a + b - d - g + 15) % 30;
    int i = c / 4;
    int k = c % 4;
    int l = (32 + 2 * e + 2 * i - h - k) % 7;
    int m = (a + 11 * h + 22 * l) / 451;
    int month = (h + l - 7 * m + 114) / 31;
    int day = ((h + l - 7 * m + 114) % 31) + 1;
    return LocalDate.of(year, month, day);
  }

  private LocalDate calculateDateOfFridays(int year, int week) {
    LocalDate firstDayOfYear = LocalDate.of(year, 1, 1);
    LocalDate firstFriday = firstDayOfYear.with(TemporalAdjusters.nextOrSame(DayOfWeek.FRIDAY));
    return firstFriday.plusWeeks(week);
  }
}

